### PR TITLE
Part of #21421: replace define with constexpr

### PR DESCRIPTION
**Pull request by janclod**
Part of #21421 

Different refactor (mostly few lines of code per replaced `define`), some unused `#define` deleted, one define (`ADD_CLAMP_BODY`) replaced with a template function.

~I realize that this PR may be too big to review at once, please let me know if I should split it.~
Since this PR is a duplicate of #21705 and it incorporates all the feedback given there, I am leaving as is.

### CODE REVIEW

**Comment by ZehMatt** [comment]
Please don't close PRs if you start over, instead force push.

**Comment by janclod** [comment]
> Please don't close PRs if you start over, instead force push.

That was my initial idea, but I failed :pensive: after a rebase, I did delete the upstream branch, this automatically closed the PR and I could not reopen. What is a good workflow, rebase and then force push? I am not familiar with force-push, can I do this after a rebase?

**Comment by ZehMatt** [comment]
> > Please don't close PRs if you start over, instead force push.
> 
> That was my initial idea, but I failed 😔 after a rebase, I did delete the upstream branch, this automatically closed the PR and I could not reopen. What is a good workflow, rebase and then force push? I am not familiar with force-push, can I do this after a rebase?

Force push just means it will overwrite the branch forcefully which means also its rewriting its history. You can always reset a branch to its base by using `git reset --hard upstream/develop` and then force push, then you can start adding commits again.

**Comment by Broxzier** [comment]
Push only after you've already made a new commit, because otherwise your branch will be the same as develop (or even behind) and GitHub will automatically close the PR.

**Comment by janclod** [comment]
Yeeey :D this time I forced pushed... Thanks for teaching me!

File: src/openrct2/util/Util.cpp [hunk 1]
```diff
@@ -502,43 +502,40 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
...
-    }                                                                                                                          \
-    else if ((value_to_add < 0) && (value < (min_cap - (value_to_add))))                                                       \
-    {                                                                                                                          \
-        value = min_cap;                                                                                                       \
-    }                                                                                                                          \
-    else                                                                                                                       \
-    {                                                                                                                          \
-        value += value_to_add;                                                                                                 \
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add, T min_cap, T max_cap)
```
[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/openrct2/util/Util.cpp:505]: I think it definitely beats the macro use. Also you are correct, we can obtain the limits with numeric_limits<T>, no need to pass them as arguments.

[the remark this answers is not in the document]
Comment (janclod) [reply src/openrct2/util/Util.cpp:505]: Happy to see the template approach appreciated :) I will improve based on feedback, until the next push ;)

**Review by tupaschoal — APPROVED** [review]
This LGTM. Should definitely be squash merged due to the confusion on some merges.

Might require a few rebases still due to network version changing (for #21850 and maybe more by the time you see it)

File: src/openrct2/util/Util.cpp [hunk 2]
```diff
@@ -502,43 +502,42 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
...
-    else if ((value_to_add < 0) && (value < (min_cap - (value_to_add))))                                                       \
-    {                                                                                                                          \
-        value = min_cap;                                                                                                       \
-    }                                                                                                                          \
-    else                                                                                                                       \
-    {                                                                                                                          \
-        value += value_to_add;                                                                                                 \
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add)
+{
+    auto max_cap = std::numeric_limits<T>::max();
```
Comment (ZehMatt) [inline src/openrct2/util/Util.cpp:507]:
```suggestion
    auto maxCap = std::numeric_limits<T>::max();
```

File: src/openrct2/util/Util.cpp [hunk 3]
```diff
@@ -502,43 +502,42 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
...
+    auto min_cap = std::numeric_limits<T>::lowest();
```
Comment (ZehMatt) [inline src/openrct2/util/Util.cpp:508]:
```suggestion
    auto minCap = std::numeric_limits<T>::lowest();
```

File: src/openrct2/util/Util.cpp [hunk 4]
```diff
@@ -502,43 +502,42 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
...
-    }                                                                                                                          \
-    else if ((value_to_add < 0) && (value < (min_cap - (value_to_add))))                                                       \
-    {                                                                                                                          \
-        value = min_cap;                                                                                                       \
-    }                                                                                                                          \
-    else                                                                                                                       \
-    {                                                                                                                          \
-        value += value_to_add;                                                                                                 \
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add)
```
Comment (ZehMatt) [inline src/openrct2/util/Util.cpp:505]:
```suggestion
template<typename T> constexpr T kAddClampBody(T value, T valueToAdd)
```

**Review by ZehMatt — CHANGES_REQUESTED** [review]
Left a few remarks, we don't do snake_case, the rest seems fine to me.

**Comment by janclod** [comment]
@ZehMatt thanks for catching the camel cases, incorporated suggestions.

Also improved further the template function `AddClamp` in `Util.h`, please have a last look, do not expect big surprises.

@tupaschoal also rebased on `develop` to with correct network versioning.

**Review by janclod — COMMENTED** [review]
Incorporated changes.

**Review by tupaschoal — APPROVED** [review]

**Comment by tupaschoal** [comment]
@Gymnasiast or @ZehMatt this looks ready to go (before another merge conflict)

**Review by ZehMatt — APPROVED** [review]
LGTM, Thanks.

**Pull request merged**